specs/_features/whisk/beacon-chain.md

```diff
@@ -238,8 +238,12 @@ def process_epoch(state: BeaconState) -> None:
 ```python
 def process_whisk_opening_proof(state: BeaconState, block: BeaconBlock) -> None:
     tracker = state.whisk_proposer_trackers[state.slot % WHISK_PROPOSER_TRACKERS_COUNT]
-    k_commitment = state.whisk_k_commitments[block.proposer_index]
-    assert IsValidWhiskOpeningProof(tracker, k_commitment, block.body.whisk_opening_proof)
+    if block.body.whisk_opening_proof == WhiskTrackerProof():
+        # no proof signals tracker is created with insecure initial 'k'
+        assert tracker.k_r_G == tracker.r_G * get_initial_whisk_k(block.proposer_index, 0)
```

Comment:
good point, I think both `get_initial_whisk_k(0)` and  `get_unique_whisk_k` may not return the correct k. `get_unique_whisk_k` returns a k that unique among the set of commitments when the validator was onboarded. Now the set of commitments is different. So it may return a different k. An option is for the validator to submit k suffixed to the zero proof

Reply:
Pushed a fix to allow the proposer to specify the nonce, while still ensuring proof of possession. This whole bootstrapping business is ugly

Reply:
Hmm, I'm not used to "nonce" and "proof of possession" being used in the context of Whisk. Can you describe a bit more what the latest changes aim to do? I also see you are digging into the guts of `whisk_opening_proof` which is an opaque type.

Reply:
`nonce` in this context refers to the second argument of the `get_initial_whisk_k` function. The output of ` get_unique_whisk_k` is dependent of which specific state it is run on. So we can't just call `get_unique_whisk_k(state) with this block state since it can produce a different result than when the validator was on-boarded into the protocol. I've added an illustrative example here https://hackmd.io/@dapplion/expensive_duty_discovery#Appendix-get_unique_whisk_k-non-determinism

So the proposer must tell protocol participants the specific nonce to use on `get_unique_whisk_k()`. A hacky way to add this data without extra block body fields is to insert it into the zero-ed proof field.
- If non-zero: check proof
- If zero (section): read nonce from it

From the POV of the protocol the proof is still an opaque type, a string of bytes to pass to the crypto black box, which if partially zero-ed conveys meaning.

Reply:
I've added the nonce as another property of the block I think it much clearer this way, @asn-d6  https://github.com/ethereum/consensus-specs/pull/3481/commits/bab555a029801963a7aa09832125b0cd353d72f8